### ISSUE TITLE
[AIR 33.1.1.698] Failed to package Android AAB with some Windows devices

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#1409]
## Problem Description
Failed to package Android AAB with some Windows devices using AIR 33.1.1.698.

It has been tested with latest AIR SDK 33.1.1.698 with different Windows 10 64-bit devices and Java 11.0.12.
With one Windows 10 machine all works fine, with others no. All settings (Java versions, adt.cfg, environment variables etc) seems to be the same.
All devices uses Java 11.0.12. (Java 8 cannot be used anymore to package AAB) (Java 17 has linked issues)
There is no such problem with AIR SDK 33.1.1.686 and below with the same Windows machines.
Didn't test it for macOS environment.

What tried (but didn't help):
- Changing Java installation path (without spaces etc).
- Set up JAVA_HOME via adt.cfg
- Set up AndroidPlatformSDK via adt.cfg

Related issues (not the same):
https://github.com/airsdk/Adobe-Runtime-Support/issues/1405
https://github.com/airsdk/Adobe-Runtime-Support/issues/1368

## Steps to Reproduce
Try to package AIR application as Android AAB bundle with Windows machine and Java 11.
Packaging command:
`adt -package -target aab -storetype pkcs12 -keystore android.p12 -storepass password -keypass password test_app.aab test_app-app.xml -platformsdk "C:/Programs/sdk/android/" -extdir libs test_app.swf`

**Actual Result:**
Packaging error:
```
NOTE: Multiple Android build-tools folders exist, choosing the latest version (30.0.3) for packaging
unexpected failure: Unable to run jarsigner: com.adobe.air.ADTException: Android App Bundler signing failed:
java.io.IOException: Unable to run jarsigner: com.adobe.air.ADTException: Android App Bundler signing failed:
at com.adobe.air.apk.AABOutputStream.jarSign(AABOutputStream.java)
at com.adobe.air.apk.AABPackager.renameOutput(AABPackager.java)
at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
at com.adobe.air.apk.AABPackager.createPackage(AABPackager.java)
at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
at com.adobe.air.ADT.run(ADT.java)
at com.adobe.air.ADT.main(ADT.java)
```

**Expected Result:**
AAB bundle will be packaged without errors.

## Known Workarounds
Use AIR SDK 33.1.1.686.

**Comment by ajwfrost** [comment]
If `jarsigner` failed, it would be good to check (a) that you have jarsigner.exe in your JDK path (with the bundle formats, we already check that it's a JDK i.e. it has `javac.exe` so it should have `jarsigner.exe` too...?) - and then (b) whether there's a problem running it from the command line. If you enable the logging, you should see the command-line being used for the jarsigner process, and this can just be copied/pasted to a normal command-line.

thanks

**Comment by marchbold** [comment]
I'm getting a similar error (under macOS though). `jarsigner` is in my java jdk (`/Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/jarsigner`)

```
*** ADT Logging at 2021/12/03 17:23:14 ***
Called by: /Users/marchbold/work/sdks/air/current/lib/adt.jar
ADT 33.1.1.698 called with: -package -target apk-debug -listen 7936 -storetype PKCS12 -keystore /Users/marchbold/Google Drive/work/certs/distriqt_air_selfsign_cert.p12 -storepass xxxxxx /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/starling.apk /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/TestDialog-app.xml -extdir /Users/marchbold/work/distriqt/extensions/supportextensions/core/source/build -extdir /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/build -extdir /Users/marchbold/work/distriqt/extensions/extensions/debug/source/build -extdir /Users/marchbold/work/distriqt/extensions/supportextensions/androidsupport/source/extensions/x-core/build -extdir /Users/marchbold/work/distriqt/extensions/supportextensions/androidsupport/source/extensions/x-appcompat/build -extdir /Users/marchbold/work/distriqt/extensions/supportextensions/androidsupport/source/extensions/x-vectordrawable/build -extdir /Users/marchbold/work/distriqt/extensions/supportextensions/androidsupport/source/extensions/kotlin/build -C /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out starling.swf -C /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/src icons
Overriding request for apk-debug, instead generating Android App Bundle
AAB Packager entry point
Version in APK = 33.1.1.698, version of ADT = 33.1.1.698
Version in APK = 33.1.1.698, version of ADT = 33.1.1.698
Version in APK = 33.1.1.698, version of ADT = 33.1.1.698
Version in APK = 33.1.1.698, version of ADT = 33.1.1.698
Gradle build where ADT is using JRE version 11.0.12
but where Gradle will be using the JDK from /Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/java
Gradle build command:  /Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/java -Xmx512m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx512m -Duser.dir="/Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/AndroidStudioProject" -cp "/Users/marchbold/work/sdks/air/current/lib/android/bin/gradle-wrapper.jar":"/Users/marchbold/work/sdks/air/current/lib/android/bin/gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:bundleRelease
AAB signer command:  /Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/jarsigner -storetype PKCS12 -keystore /Users/marchbold/Google Drive/work/certs/distriqt_air_selfsign_cert.p12 -storepass xxxx /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/AndroidStudioProject/app/build/outputs/bundle/release/app-release.aab 1
..java.base/java.lang.Thread.getStackTrace(Thread.java:1602)
..com.adobe.air.DebugFile.printStack(DebugFile.java)
..com.adobe.air.ADTException.<init>(ADTException.java)
..com.adobe.air.apk.AABOutputStream.jarSign(AABOutputStream.java)
..com.adobe.air.apk.AABPackager.renameOutput(AABPackager.java)
..com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
..com.adobe.air.apk.AABPackager.createPackage(AABPackager.java)
..com.adobe.air.ADT.parseArgsAndGo(ADT.java)
..com.adobe.air.ADT.run(ADT.java)
..com.adobe.air.ADT.main(ADT.java)
Stack trace:
Unexpected failure: Unable to run jarsigner: com.adobe.air.ADTException: Android App Bundler signing failed: Only one alias can be specified
```

Is the space character in the certificate path an issue?

**Comment by marchbold** [comment]
I just tried to run the command in terminal, and the space character definitely seems to be an issue. 

```
/Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/jarsigner -storetype PKCS12 -keystore /Users/marchbold/Google Drive/work/certs/distriqt_air_selfsign_cert.p12 -storepass xxxxx /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/AndroidStudioProject/app/build/outputs/bundle/release/app-release.aab 1
Only one alias can be specified
```

If I quote it:

```
/Library/Java/JavaVirtualMachines/jdk-11.0.13.jdk/Contents/Home/bin/jarsigner -storetype PKCS12 -keystore "/Users/marchbold/Google Drive/work/certs/distriqt_air_selfsign_cert.p12" -storepass xxxxx /Users/marchbold/work/distriqt/extensions/extensions/dialog/source/example/starling/out/AndroidStudioProject/app/build/outputs/bundle/release/app-release.aab 1
jar signed.
```

**Comment by ajwfrost** [comment]
Yes, sorry .. I really thought we had tested with space characters in the path, as we've been caught out by this before! Will get it checked/updated..

thanks

**Comment by itlancer** [comment]
@ajwfrost
Yes, seems that problem with escaping (quoting) paths. If some of paths contains spaces or something like `../../cert.p12` it will fail. Please fix it ASAP.

**Comment by itlancer** [comment]
@ajwfrost
Latest patch https://github.com/airsdk/Adobe-Runtime-Support/issues/1412#issuecomment-985462226 fixed issue with spaces in paths but didn't fix `../../cert.p12` usage.
Full adt.log:
```
Gradle build where ADT is using JRE version 11.0.12
but where Gradle will be using the JDK from C:\Programs\Java\jdk-11.0.12\bin\java.exe
Gradle build command: C:\Programs\Java\jdk-11.0.12\bin\java.exe -Xmx2048m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx2048m -Duser.dir="C:\Users\username\AppData\Local\Temp\80572126-20b0-4152-a151-abdae1047372" -cp "C:\Programs\SDK\air\33\lib\android\bin\gradle-wrapper.jar";"C:\Programs\SDK\air\33\lib\android\bin\gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:bundleRelease
AAB signer command: C:\Programs\Java\jdk-11.0.12\bin\jarsigner.exe -storetype pkcs12 -keystore ../../cert/cert.p12 -storepass xxxx -keypass xxxx C:\Users\username\AppData\Local\Temp\80572126-20b0-4152-a151-abdae1047372/app/build/outputs/bundle/release/app-release.aab 1
..java.base/java.lang.Thread.getStackTrace(Thread.java:1606)
..com.adobe.air.DebugFile.printStack(DebugFile.java)
..com.adobe.air.ADTException.<init>(ADTException.java)
..com.adobe.air.apk.AABOutputStream.jarSign(AABOutputStream.java)
..com.adobe.air.apk.AABPackager.renameOutput(AABPackager.java)
..com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
..com.adobe.air.apk.AABPackager.createPackage(AABPackager.java)
..com.adobe.air.ADT.parseArgsAndGo(ADT.java)
..com.adobe.air.ADT.run(ADT.java)
..com.adobe.air.ADT.main(ADT.java)
Stack trace:
Unexpected failure: Unable to run jarsigner: com.adobe.air.ADTException: Android App Bundler signing failed:
```

**Comment by ajwfrost** [comment]
@itlancer  good spot, thanks ... will sort out the relative addressing too...

**Comment by ajwfrost** [comment]
Hi
Please find an updated version which should handle working folders and relative paths:
https://transfer.harman.com/message/X33u6l0ycLmvcm8CtvCxZi
thanks

**Comment by itlancer** [comment]
@ajwfrost
Now it fixed with latest patch. Thank you!
Waiting for a new release to close this issue.

**Comment by itlancer** [comment]
Fixed with latest AIR 33.1.1.713.
Thanks!
